synapse/lib/cell.py

```diff
@@ -4031,3 +4047,219 @@ async def _runReloadFunc(self, name, func):
             return s_common.retnexc(e)
         logger.debug(f'Completed cell reload system {name}')
         return (True, ret)
+
+    async def addUserApiKey(self, useriden, name, duration=None):
+
+        user = await self.auth.reqUser(useriden)
+        iden, token, shadow = await s_passwd.generateApiKey()
+        now = s_common.now()
+        kdef = {
+            'iden': iden,
+            'name': name,
+            'user': user.iden,
+            'created': now,
+            'modified': now,
+            'shadow': shadow,
+            'duration': None,
+            'expref': None,
+        }
+
+        if duration:
+            kdef['expref'] = now
+            kdef['duration'] = duration
+
+        kdef = s_schemas.reqValidUserApiKeyDef(kdef)
+
+        await self._push('user:apikey:add', kdef)
+
+        logger.info(f'Created HTTP API Key {iden} for {user.name}, {name=}',
+                    extra=await self.getLogExtra(target_user=user.iden, target_username=user.name, iden=iden,
+                                                 status='MODIFY'))
+
+        kdef.pop('shadow')
+        return token, kdef
+
+    @s_nexus.Pusher.onPush('user:apikey:add')
+    async def _genUserApiKey(self, kdef):
+        iden = kdef.get('iden').encode('utf-8')
+        user = kdef.get('user').encode('utf-8')
+        self.slab.put(iden, user, db=self.apikeydb)
+        lkey = user + b'cell:apikey' + iden
+        self.slab.put(lkey, s_msgpack.en(kdef), db=self.usermetadb)
+
+    async def getUserApiKey(self, iden):
+        '''
+        Get a user API key via iden.
+
+        Args:
+            iden (str): The key iden.
+
+        Returns:
+            dict: The key dictionary; or none.
+        '''
+        user = self.slab.get(iden.encode('utf-8'), db=self.apikeydb)
+        if user is None:
+            return None
+        lkey = user + b'cell:apikey' + iden.encode('utf-8')
+        buf = self.slab.get(lkey, db=self.usermetadb)  # None here would be a nexus state inconsistency
+        kdef = s_msgpack.un(buf)  # This includes the shadow key
+        return kdef
+
+    async def listUserApiKeys(self, user=None):
```

Comment:
I think we can safely make this API require user to simplify the use case below, no? Maybe deconflate and make a `getApiKeys()`. Without user specified, it should def be a generator :)